File: document/JAVA/sort/java/MergeSort.java
/*
Merge Sort
- Stable: Yes
- In-place: No
- Time: O(n log n)
- Space: O(n)
*/
public class MergeSort {
    static void printArr(int[] a) {
        for (int i = 0; i < a.length; i++) {
            if (i > 0) System.out.print(" ");
            System.out.print(a[i]);
        }
        System.out.println();
    }
    private static void merge(int[] a, int l, int m, int r) {
        int n1 = m - l + 1, n2 = r - m;
        int[] L = new int[n1];
        int[] R = new int[n2];
        for (int i = 0; i < n1; i++) L[i] = a[l + i];
        for (int j = 0; j < n2; j++) R[j] = a[m + 1 + j];
        int i = 0, j = 0, k = l;
        while (i < n1 && j < n2) {
            if (L[i] <= R[j]) a[k++] = L[i++];
            else a[k++] = R[j++];
        }
        while (i < n1) a[k++] = L[i++];
        while (j < n2) a[k++] = R[j++];
    }
    private static void sortRec(int[] a, int l, int r) {
        if (l >= r) return;
        int m = l + (r - l) / 2;
        sortRec(a, l, m);
        sortRec(a, m + 1, r);
        merge(a, l, m, r);
    }
    public static void sort(int[] a) { sortRec(a, 0, a.length - 1); }
    public static void main(String[] args) {
        int[] a = {38,27,43,3,9,82,10};
        sort(a);
        printArr(a);
    }
}

File: document/JAVA/sort/python/merge_sort.py
"""
Merge Sort
- Stable: Yes
- In-place: No (uses extra memory)
- Time: O(n log n)
- Space: O(n)
"""

def merge_sort(arr):
    # Recursive divide-and-conquer
    if len(arr) <= 1:
        return arr
    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])
    return _merge(left, right)


def _merge(left, right):
    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:  # <= keeps stability
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result

if __name__ == "__main__":
    data = [38, 27, 43, 3, 9, 82, 10]
    print("before:", data)
    print("after :", merge_sort(data))
